### DISCUSSION TITLE
Adding support to a Hue bulb, how to trigger "some actions"?

### DISCUSSION THREAD

**Discussion by theNorr** [Koenkk/zigbee2mqtt#15581] [Devices]
I'm trying to add a Philips Hue Bulb but when I get to the part "Once finished, restart Zigbee2MQTT and trigger some actions on the device" I get stuck. 
How the heck am I supposed to trigger "some actions" on a bulb without having it connected to a hub?

It's regarding a Philips Hue White A60 E27, Ean: 8719514288232.

**Comment by danieledwardgeorgehitchcock** [comment]
Hi,

So can you take a step back and describe what you have done so far? Have you created an External Converter? if so, could you post it up here?

**Comment by theNorr** [comment]
Yes, I've copied the one in the official guide and changed this info:
```
    zigbeeModel: ['8719514288232'], // The model ID from: Device with modelID 'lumi.sens' is not supported.
    model: '8719514288232', // Vendor model number, look on the device for a model number
    vendor: 'Philips', // Vendor of the device (only used for documentation and startup logging)
    description: 'Philips Hue White A60 E27', // Description of the device, copy from vendor site. (only used for documentation and startup logging)
```

**Comment by danieledwardgeorgehitchcock** [comment]
Ok, you will need to create some references to some files such as fromZigbee, toZigbee, etc - this should also be in the guide.

From that point on, the easiest way to construct a converter suitable for your device is to look at the zigbee-herdsman-converters and find a suitable device match that you could base yours off of - why re-invent the wheel? :-)

I'd suggest this one:
https://github.com/Koenkk/zigbee-herdsman-converters/blob/4622442a38af8e7e14a4dd9e4a6aac526bd04c25/devices/philips.js#L841

hue is a bit more complex for an External Converter as you will also need to add the following to your file:

```javascript
const extendDontUse = require('../lib/extend');
const extend = {switch: extendDontUse.switch};

const hueExtend = {
    light_onoff_brightness: (options={}) => ({
        ...extendDontUse.light_onoff_brightness(options),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness(options).toZigbee.concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
    light_onoff_brightness_colortemp: (options={}) => ({
        ...extendDontUse.light_onoff_brightness_colortemp(options),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness_colortemp(options).toZigbee
            .concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
    light_onoff_brightness_color: (options={}) => ({
        ...extendDontUse.light_onoff_brightness_color({supportsHS: true, ...options}),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness_color({supportsHS: true, ...options}).toZigbee
            .concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
    light_onoff_brightness_colortemp_color: (options={}) => ({
        ...extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options}),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options})
            .toZigbee.concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
    light_onoff_brightness_colortemp_color_gradient: (options={}) => ({
        ...extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, noConfigure: true, ...options}),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        configure: async (device, coordinatorEndpoint, logger) => {
            await extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options})
                .configure(device, coordinatorEndpoint, logger);
            for (const ep of device.endpoints) {
                await ep.bind('manuSpecificPhilips2', coordinatorEndpoint);
            }
        },
        exposes: extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options}).exposes.concat([
            // gradient_scene is deprecated, use gradient instead
            exposes.enum('gradient_scene', ea.SET, Object.keys(gradientScenes)),
            exposes.list('gradient', ea.ALL, exposes.text('hex', 'Color in RGB HEX format (eg #663399)'))
                .withLengthMin(1)
                .withLengthMax(9)
                .withDescription('List of RGB HEX colors'),
        ]),
        fromZigbee: extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options}).fromZigbee.concat(
            [fzLocal.gradient({reverse: true})]),
        toZigbee: extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options}).toZigbee.concat(
            [tz.hue_power_on_behavior, tz.hue_power_on_error, tzLocal.gradient_scene, tzLocal.gradient({reverse: true})]),
    }),
};
```

**Reply by theNorr** [reply]
Hmm, according to the official guide I should trigger "some actions" and read the logs to be able to determine what I need to add to the fromZigbee and the toZigbee.
I just don't get how to trigger those actions.
I get that if the device has one or more buttons, this can be done, but unless it does I don't know how to.
I guess I might be able to use cluster and write to the device and see what it returns but I don't know what Cluster and attributes to use, seems like a lot to go through all of them...

I was thinking about looking through the others but couldn't really decide which to use so kind of decided that if I'm going to try to add official support for it, I should do it the "right way" the first time, to learn and understand...

**Comment by danieledwardgeorgehitchcock** [comment]
If you like, I could add the support for you to the project - this will still give you some time to make your own external converter...

**Reply by theNorr** [reply]
I'd like that!

**Comment by danieledwardgeorgehitchcock** [comment]
See here Koenkk/zigbee-herdsman-converters#5160

Sometimes it is easier to use an existing device as a template and submit a pull request, than it is to create an External Converter.

Once it is merged, it is usually available in the Dev build of Z2M within 24 hours

**Reply by theNorr** [reply]
Thank you!
How did you know that you should use "hueExtend.light_onoff_brightness_colortemp({colorTempRange: [153, 454]})"?
Did you compare specs of the devices?

**Reply by danieledwardgeorgehitchcock** [reply]
I didn't use that - I used `hueExtend.light_onoff_brightness()` My addition is just the green section [here](https://github.com/Koenkk/zigbee-herdsman-converters/pull/5160/files)

**Reply by theNorr** [reply]
Ah, sorry, looked at the wrong lines.

**Comment by danieledwardgeorgehitchcock** [comment]
I'd still recommend you play around with trying to support the device via an external converter while you are waiting for the merge.

Regarding the process, and to answer your original question, a trigger of an action is usually only required as you say for devices that have an external method of control.

When trying to add a device such as a bulb, the best course of action is to enable debug logging in Z2M, add the device and watch some of the logs come through - Z2M will raise a warning saying `no converter is available for x` - it would then be a case of finding or creating an appropriate converter to handle that event.

**Reply by theNorr** [reply]
Alright, I'll give it a try! Thank you!

**Comment by theNorr** [comment]
If I use the extend.. you added, I don't need the 'fromZigbee' and 'toZigbee', right?

**Comment by theNorr** [comment]
I get this error message when trying to use "my js-file":

`MQTT publish: topic 'zigbee2mqtt/bridge/response/device/configure', payload '{"data":{"id":"0x001788010c34ddfb"},"error":"Device '0x001788010c34ddfb' cannot be configured","status":"error","transaction":"rbv35-1"}'`

I've added the code you posted above, and the extend-line you added in you PR.

**Reply by danieledwardgeorgehitchcock** [reply]
could you post up the contents of your converter file?

**Reply by theNorr** [reply]
```const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const extendDontUse = require('../lib/extend');
const extend = {switch: extendDontUse.switch};

const hueExtend = {
    light_onoff_brightness: (options={}) => ({
        ...extendDontUse.light_onoff_brightness(options),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness(options).toZigbee.concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
    light_onoff_brightness_colortemp: (options={}) => ({
        ...extendDontUse.light_onoff_brightness_colortemp(options),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness_colortemp(options).toZigbee
            .concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
    light_onoff_brightness_color: (options={}) => ({
        ...extendDontUse.light_onoff_brightness_color({supportsHS: true, ...options}),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness_color({supportsHS: true, ...options}).toZigbee
            .concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
    light_onoff_brightness_colortemp_color: (options={}) => ({
        ...extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options}),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options})
            .toZigbee.concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
    light_onoff_brightness_colortemp_color_gradient: (options={}) => ({
        ...extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, noConfigure: true, ...options}),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        configure: async (device, coordinatorEndpoint, logger) => {
            await extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options})
                .configure(device, coordinatorEndpoint, logger);
            for (const ep of device.endpoints) {
                await ep.bind('manuSpecificPhilips2', coordinatorEndpoint);
            }
        },
        exposes: extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options}).exposes.concat([
            // gradient_scene is deprecated, use gradient instead
            exposes.enum('gradient_scene', ea.SET, Object.keys(gradientScenes)),
            exposes.list('gradient', ea.ALL, exposes.text('hex', 'Color in RGB HEX format (eg #663399)'))
                .withLengthMin(1)
                .withLengthMax(9)
                .withDescription('List of RGB HEX colors'),
        ]),
        fromZigbee: extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options}).fromZigbee.concat(
            [fzLocal.gradient({reverse: true})]),
        toZigbee: extendDontUse.light_onoff_brightness_colortemp_color({supportsHS: true, ...options}).toZigbee.concat(
            [tz.hue_power_on_behavior, tz.hue_power_on_error, tzLocal.gradient_scene, tzLocal.gradient({reverse: true})]),
    }),
};

const definition = {
    zigbeeModel: ['8719514288232'], // The model ID from: Device with modelID 'lumi.sens' is not supported.
    model: '8719514288232', // Vendor model number, look on the device for a model number
    vendor: 'Philips', // Vendor of the device (only used for documentation and startup logging)
    description: 'Philips Hue White A60 E27', // Description of the device, copy from vendor site. (only used for documentation and startup logging)
    extend: hueExtend.light_onoff_brightness(),
};

module.exports = definition;```

**Reply by theNorr** [reply]
I think I see it...

**Reply by danieledwardgeorgehitchcock** [reply]
Sorry, change the relative reference for `extendDontUse` to an absolute `const extendDontUse = require('../lib/extend');` -> `const extendDontUse = require('zigbee-herdsman-converters/lib/extend');`

**Reply by danieledwardgeorgehitchcock** [reply]
You'll also need a reference to the ota library too:

`const ota = require('zigbee-herdsman-converters/lib/ota');`

**Reply by theNorr** [reply]
Still doesn't work...

Do I need all of these?

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const extendDontUse = require('zigbee-herdsman-converters/lib/extend');
const extend = {switch: extendDontUse.switch};
const ota = require('zigbee-herdsman-converters/lib/ota');
```

**Reply by theNorr** [reply]
I have two entries for `const extend` for example, is that really necessary? Can that be the cause?

**Reply by danieledwardgeorgehitchcock** [reply]
Just noticed that. As it stands, you can get rid of both entries alongside:

```
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const e = exposes.presets;
```

**Reply by theNorr** [reply]
Still the same error when trying to reconfigure the device...

**Reply by theNorr** [reply]
Do I need the 'fromZigbee' and the 'toZigbee'?
I removed them from the `const definition` and added `extend: hueExtend.light_onoff_brightness()`, instead, can that be causing the issues?

**Reply by danieledwardgeorgehitchcock** [reply]
Let's try stripping back the `hueExtend` to just the basics. Try this:

```javascript
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const extendDontUse = require('zigbee-herdsman-converters/lib/extend');
const ota = require('zigbee-herdsman-converters/lib/ota');

const hueExtend = {
    light_onoff_brightness: (options={}) => ({
        ...extendDontUse.light_onoff_brightness(options),
        ota: ota.zigbeeOTA,
        meta: {turnsOffAtBrightness1: true},
        toZigbee: extendDontUse.light_onoff_brightness(options).toZigbee.concat([tz.hue_power_on_behavior, tz.hue_power_on_error]),
    }),
};

const definition = {
    zigbeeModel: ['8719514288232'],
    model: '929002469216',
    vendor: 'Philips',
    description: 'Hue white E27 1100lm with Bluetooth',
    extend: hueExtend.light_onoff_brightness(),
};

module.exports = definition;
```

**Reply by danieledwardgeorgehitchcock** [reply]
I think there is an unneeded and undefined reference to a `fzLocal` converter in the `hueExtend` declaration which is causing the problems

**Reply by theNorr** [reply]
Same error. Can it be something with my addon-config for the external converter?
Where should the file be stored when using the Home Assistant addon, and how should it be referenced under external_converter in the Zigbee2mqtt configuration?

**Reply by danieledwardgeorgehitchcock** [reply]
I made a few tweaks to the above code snippet - can you check that you picked up the latest version?

**Reply by theNorr** [reply]
I have.
I'm leaning towards the reference to the file in the config file...
I think I might have to move the file, or change the directory in the addon config...

**Reply by theNorr** [reply]
Pretty sure that is it...
https://www.reddit.com/r/homeassistant/comments/rwg1ly/comment/hrcd1no/?utm_source=share&utm_medium=web2x&context=3

**Reply by danieledwardgeorgehitchcock** [reply]
I'd have to leave that one with you - the code above should work and my Z2M dev environment is outside of HA.

**Reply by theNorr** [reply]
Alright, I used the terminal and added the file to the correct directory, still doesn't work.

**Reply by danieledwardgeorgehitchcock** [reply]
Realised I had missed out the OTA reference again! try the above file once more...

**Reply by theNorr** [reply]
Nope. Still doesn't work.

**Reply by theNorr** [reply]
I've managed to confirm that the external converter has been loaded.
Will try your file again in a minute.

**Reply by theNorr** [reply]
Doesn't work.
Might have to admit defeat.
If you come up with something, let me know. I'll leave it for now.

**Comment by theNorr** [comment]
Back at it...

So I've gone through the code of similar to compare, and it seems correct, and the converter file is loaded correctly.
When I try to re-configure the device I get the following log:
Received MQTT message on 'zigbee2mqtt/bridge/request/options' with data '{"options":{"external_converters":["8719514288232.js"]},"transaction":"4kzhp-2"}'
So, looks like it is working, just that I don't get a response from the device...
Have you come across this before?

**Comment by danieledwardgeorgehitchcock** [comment]
My PR has now been merged so you should be able to switch to the dev build in the next day or so and the device should show as supported

**Reply by theNorr** [reply]
Thank you! When you say switch to dev build, you mean remove the external converter file and reference?
